modules/programs/wlogout.nix

```
@@ -0,0 +1,160 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib)
+    all filterAttrs hasAttr isStorePath literalExpression optionalAttrs types;
+  inherit (lib.options) mkEnableOption mkOption;
+  inherit (lib.modules) mkIf mkMerge;
+  inherit (builtins) toJSON concatStringsSep;
+
+  cfg = config.programs.wlogout;
+
+  jsonFormat = pkgs.formats.json { };
+
+  mkMargin = name:
+    mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      example = 10;
+      description = "Margin value without unit.";
+    };
+
+  wlogoutLayoutConfig = with types;
+    submodule {
+      freeformType = jsonFormat.type;
+
+      options = {
+        label = mkOption {
+          type = str;
+          default = "";
+          example = "shutdown";
+          description = "CSS label of button.";
+        };
+
+        action = mkOption {
+          type = str;
+          default = "";
+          example = "systemctl poweroff";
+          description = "Command to execute when clicked.";
+        };
+
+        text = mkOption {
+          type = str;
+          default = "";
+          example = "Shutdown";
+          description = "Text displayed on button.";
+        };
+
+        keybind = mkOption {
+          type = str;
+          default = "";
+          example = "s";
+          description = "Keyboard character to trigger this action.";
+        };
+
+        height = mkOption {
+          type = nullOr (numbers.between 0 1);
+          default = null;
+          example = 0.5;
+          description = "Relative height of tile.";
+        };
+
+        width = mkOption {
+          type = nullOr (numbers.between 0 1);
+          default = null;
+          example = 0.5;
+          description = "Relative width of tile.";
+        };
+
+        circular = mkOption {
+          type = nullOr bool;
+          default = null;
+          example = true;
+          description = "Make button circular.";
+        };
+      };
+    };
+in {
+  meta.maintainers = [ lib.maintainers.Scrumplex ];
+  options.programs.wlogout = with lib.types; {
+    enable = mkEnableOption "wlogout";
+
+    package = mkOption {
+      type = package;
+      default = pkgs.wlogout;
+      defaultText = literalExpression "pkgs.wlogout";
+      description = ''
+        wlogout package to use. Set to <code>null</code> to use the default package.
+      '';
+    };
```

```suggestion
    package = mkPackageOption pkgs "wlogout" { };
```

Just to be pedantic, the bit about `null` is true only if the user sets the `package` value with a priority less than `mkOptionDefault`.
https://github.com/NixOS/nixpkgs/blob/bea4062a62e18615421e8f4ce7bc339529f1c106/lib/modules.nix#L904-L922

---

No, it's not true at all. Setting the option to `null` is a type error because `null` is not of type `package`. To use the default package, you just don't set the option.

---

yes, as well as missing the `nullOr` :)
Thanks for correcting me